src/main.rs

```diff
@@ -26,30 +28,56 @@ fn get_default_device() -> Option<&'static str> {
         .map(|&str| str)
 }
 
+#[derive(Debug, Clone)]
 struct ProgramState {
     controllers: HashSet<controller::Controller>,
-    nodes: HashSet<Node>,
-    nodes_map: HashMap<controller::Controller, Vec<Node>>,
+    nodes: BTreeSet<Node>,
+    nodes_map: HashMap<controller::Controller, BTreeSet<Node>>,
+    value_ids: BTreeSet<ValueID>,
```

Comment:
I think we'll need a node -> value_ids map as well. I don't know if it's "as well" or "instead". I'd say "instead" is most likely but we can figure this out later when we build the adapter.